library/core/src/alloc/global.rs

```diff
@@ -20,24 +20,55 @@ use crate::ptr;
 ///
 /// # Example
 ///
-/// ```no_run
-/// use std::alloc::{GlobalAlloc, Layout, alloc};
+/// ```
+/// use std::alloc::{GlobalAlloc, Layout};
+/// use std::cell::UnsafeCell;
 /// use std::ptr::null_mut;
+/// use std::sync::atomic::{AtomicUsize, Ordering::{Acquire, SeqCst}};
 ///
-/// struct MyAllocator;
-///
-/// unsafe impl GlobalAlloc for MyAllocator {
-///     unsafe fn alloc(&self, _layout: Layout) -> *mut u8 { null_mut() }
-///     unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+/// const ARENA: usize = 100 * 1024;
+/// struct SimpleAllocator {
+///     arena: UnsafeCell<[u8; ARENA]>,
+///     remaining: AtomicUsize, // we allocate from the top, counting down
 /// }
 ///
 /// #[global_allocator]
-/// static A: MyAllocator = MyAllocator;
+/// static ALLOCATOR: SimpleAllocator = SimpleAllocator {
+///     arena: UnsafeCell::new([0x55; ARENA]),
+///     remaining: AtomicUsize::new(ARENA),
+/// };
 ///
-/// fn main() {
-///     unsafe {
-///         assert!(alloc(Layout::new::<u32>()).is_null())
+/// unsafe impl Sync for SimpleAllocator { }
+///
+/// unsafe impl GlobalAlloc for SimpleAllocator {
+///     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+///         let size = layout.size();
+///
+///         // Both align==0 and align not power of 2 will mangle remaining (or panic), causing UB.
+///         // This is allowed by the `GlobalAllocator` and `Layout` contracts.
+///         let align_mask_to_round_down = !(layout.align() - 1);
+///
+///         let mut allocated = 0;
+///         if self.remaining.fetch_update(SeqCst, SeqCst, |mut remaining| {
+///             if size > remaining {
+///                 return None
+///             }
+///             remaining -= size;
+///             remaining &= align_mask_to_round_down;
```

Comment:
There are alignment + size combinations where this is incorrect - too big alignment and we end up with an allocation that overlaps with a previous one

Reply:
Thanks for the detailed review.  I appreciate your critical eye.

> In addition, nothing accounts for the alignment of `self.arena.get()`.

You are entirely right.  I will add `#[repr(C, align( ))]`.

> There are alignment + size combinations where this is incorrect unfortunately - too big alignment and we end up with an allocation is misaligned

This is a consequence of the former.  The construction ensures that the offset is a multiple of the alignment.  If the alignment requested is greater than the alignment of the buffer, this goes wrong.  That can only happen if the alignment is greater than the size which is unusual, but permitted by the API.

> Even as an example, I think the allocator must check all its preconditions and fail the allocation otherwise.

Yes.  Absolutely.

Reply:
Fixed these issues I think.